drift/lib/src/runtime/query_builder/expressions/case_when.dart

```diff
@@ -2,35 +2,113 @@ import 'package:meta/meta.dart';
+///
+/// [Multiple return values GitHub issue](https://github.com/dart-lang/language/issues/68)
+class CaseWhen<T extends Object, R extends Object> {
+  /// Expression to use after `WHEN`
+  final Expression<T> when;
+
+  /// Expression to use after `THEN`
+  final Expression<R> then;
+
+  /// Creates a single case with expression for when and expression for then.
+  const CaseWhen(this.when, {required this.then});
+}
+
+/// `CASE WHEN` expression in sqlite.
+///
+/// This expression evaluates to the first value in [BaseCaseWhenExpression.orderedCases]
+/// for which its key evaluates to `true`.
+///
+/// Supports `CASE WHEN` expression with base expression and without base expression.
+///
+/// For internal use. Must be used via [CaseWhenExpression] and [CaseWhenExpressionWithBase]
+///
+/// https://www.sqlite.org/lang_expr.html#the_case_expression
 @internal
-class CaseWhenExpression<T extends Object> extends Expression<T> {
-  /// The optional base expression. If it's set, the keys in [whenThen] will be
-  /// compared to this expression.
+abstract class BaseCaseWhenExpression<T extends Object, R extends Object>
+    extends Expression<R> {
+  /// The optional base expression.
+  ///
+  /// If it is set, the keys in [orderedCases] will be compared to this expression.
   final Expression? base;
 
-  /// The when entries for this expression. This expression will evaluate to the
-  /// value of the entry with a matching key.
-  final List<MapEntry<Expression, Expression>> whenThen;
+  /// The when-then entries for this expression.
+  ///
+  /// This expression will evaluate to the value of the entry with a matching key.
+  final List<CaseWhen<T, R>> orderedCases;
 
-  /// The expression to use if no entry in [whenThen] matched.
-  final Expression<T>? orElse;
+  /// The expression to use if no entry in [orderedCases] matched.
+  final Expression<R>? orElse;
 
   /// Creates a `CASE WHEN` expression from the independent components.
-  CaseWhenExpression(this.base, this.whenThen, this.orElse);
+  ///
+  /// If [orderedCases] list is empty - throws [ArgumentError]
+  BaseCaseWhenExpression._(this.base, this.orderedCases, this.orElse) {
+    if (orderedCases.isEmpty) {
+      throw ArgumentError.value(
+        orderedCases,
+        'orderedCases',
+        'Must not be empty',
+      );
+    }
+  }
 
   @override
   void writeInto(GenerationContext context) {
-    context.buffer.write('CASE ');
-    base?.writeInto(context);
+    context.buffer.write('CASE');
+
+    if (base != null) {
+      context.buffer.write(' ');
+      base!.writeInto(context);
+    }
 
-    for (final entry in whenThen) {
+    for (final specificCase in orderedCases) {
       context.buffer.write(' WHEN ');
-      entry.key.writeInto(context);
+      specificCase.when.writeInto(context);
       context.buffer.write(' THEN ');
-      entry.value.writeInto(context);
+      specificCase.then.writeInto(context);
```

Comment:
https://dart.dev/guides/language/effective-dart/design#consider-making-the-code-read-like-a-sentence